flow/layers/display_list_raster_cache_item.cc

```diff
@@ -0,0 +1,177 @@
+  }
+
+  if (is_complex) {
+    // The caller seems to have extra information about the display list and
+    // thinks the display list is always worth rasterizing.
+    return true;
+  }
+
+  unsigned int complexity_score = complexity_calculator->Compute(display_list);
+  return complexity_calculator->ShouldBeCached(complexity_score);
+}
+
+DisplayListRasterCacheItem::DisplayListRasterCacheItem(
+    DisplayList* display_list,
+    const SkPoint& offset,
+    bool is_complex,
+    bool will_change)
+    : RasterCacheItem(RasterCacheKeyID(display_list->unique_id(),
+                                       RasterCacheKeyType::kDisplayList),
+                      CacheState::kCurrent),
+      display_list_(display_list),
+      offset_(offset),
+      is_complex_(is_complex),
+      will_change_(will_change) {}
+
+std::unique_ptr<DisplayListRasterCacheItem> DisplayListRasterCacheItem::Make(
+    DisplayList* display_list,
+    const SkPoint& offset,
+    bool is_complex,
+    bool will_change) {
+  return std::make_unique<DisplayListRasterCacheItem>(display_list, offset,
+                                                      is_complex, will_change);
+}
+
+void DisplayListRasterCacheItem::PrerollSetup(PrerollContext* context,
+                                              const SkMatrix& matrix) {
+  cache_state_ = CacheState::kNone;
+  DisplayListComplexityCalculator* complexity_calculator =
+      context->gr_context ? DisplayListComplexityCalculator::GetForBackend(
+                                context->gr_context->backend())
+                          : DisplayListComplexityCalculator::GetForSoftware();
+
+  if (!IsDisplayListWorthRasterizing(display_list_, will_change_, is_complex_,
+                                     complexity_calculator)) {
+    // We only deal with display lists that are worthy of rasterization.
+    return;
+  }
+
+  transformation_matrix_ = matrix;
+  transformation_matrix_.preTranslate(offset_.x(), offset_.y());
+
+  if (!transformation_matrix_.invert(nullptr)) {
+    // The matrix was singular. No point in going further.
+    return;
+  }
+
+  if (context->raster_cached_entries && context->raster_cache) {
+    context->raster_cached_entries->push_back(this);
+    cache_state_ = CacheState::kCurrent;
+  }
+  return;
+}
+
+void DisplayListRasterCacheItem::PrerollFinalize(PrerollContext* context,
+                                                 const SkMatrix& matrix) {
+  if (cache_state_ == CacheState::kNone || !context->raster_cache ||
+      !context->raster_cached_entries) {
+    return;
+  }
+  auto* raster_cache = context->raster_cache;
+  SkRect bounds = display_list_->bounds().makeOffset(offset_.x(), offset_.y());
+  // We must to create an entry whenever if the react is intersect.
+  // if the rect is intersect we will get the entry access_count to confirm if
+  // it great than the threshold. Otherwise we only increase the entry
+  // access_count.
+  if (context->cull_rect.intersect(bounds)) {
```

Comment:
Shouldn't this be "intersects"? Is this a bug? I'm curious why it never showed up anywhere. Perhaps because we don't use this cull_rect for much during preroll?